# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/MediaStreamTrack.cpp
# Commit: c54cb3c10992
# Full Hash: c54cb3c109922bc15fb6d0ca67e5f2a05980c1a5
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2019-07-31 15:46:43
# Regressor Bug: 1493613
# File Overlap Count: 2
# Description:
#   Bug 1493613 - Move MediaStream control from DOMMediaStream to MediaStreamTrack. r=padenot
#   
#   This is inherently large, because modifying these bits of DOMMediaStream and
#   MediaStreamTrack affects all consumers and producers of all DOMMediaStreams and
#   MediaStreamTracks.
# ==============================================================================

diff -r 0b03dd9d20ac -r c54cb3c10992 dom/media/MediaStreamTrack.cpp
--- a/dom/media/MediaStreamTrack.cpp	Wed Jul 31 09:35:06 2019 +0000
+++ b/dom/media/MediaStreamTrack.cpp	Wed Jul 31 07:58:17 2019 +0000
@@ -139,7 +139,7 @@
   WeakPtr<MediaStreamTrack> mTrack;
 };
 
-class TrackSink : public MediaStreamTrackSource::Sink {
+class MediaStreamTrack::TrackSink : public MediaStreamTrackSource::Sink {
  public:
   explicit TrackSink(MediaStreamTrack* aTrack) : mTrack(aTrack) {}
 
@@ -168,27 +168,37 @@
     }
   }
 
+  void OverrideEnded() override {
+    if (mTrack) {
+      mTrack->OverrideEnded();
+    }
+  }
+
  private:
   WeakPtr<MediaStreamTrack> mTrack;
 };
 
-MediaStreamTrack::MediaStreamTrack(DOMMediaStream* aStream, TrackID aTrackID,
-                                   TrackID aInputTrackID,
+MediaStreamTrack::MediaStreamTrack(nsPIDOMWindowInner* aWindow,
+                                   MediaStream* aInputStream, TrackID aTrackID,
                                    MediaStreamTrackSource* aSource,
                                    const MediaTrackConstraints& aConstraints)
-    : mOwningStream(aStream),
+    : mWindow(aWindow),
+      mInputStream(aInputStream),
+      mStream(mInputStream ? mInputStream->Graph()->CreateTrackUnionStream()
+                           : nullptr),
+      mPort(mStream ? mStream->AllocateInputPort(mInputStream) : nullptr),
       mTrackID(aTrackID),
-      mInputTrackID(aInputTrackID),
       mSource(aSource),
       mSink(MakeUnique<TrackSink>(this)),
       mPrincipal(aSource->GetPrincipal()),
-      mReadyState(MediaStreamTrackState::Live),
+      mReadyState(mStream ? MediaStreamTrackState::Live
+                          : MediaStreamTrackState::Ended),
       mEnabled(true),
       mMuted(false),
       mConstraints(aConstraints) {
-  GetSource().RegisterSink(mSink.get());
+  if (!Ended()) {
+    GetSource().RegisterSink(mSink.get());
 
-  if (GetOwnedStream()) {
     mMSGListener = new MSGListener(this);
     AddListener(mMSGListener);
   }
@@ -211,16 +221,7 @@
 MediaStreamTrack::~MediaStreamTrack() { Destroy(); }
 
 void MediaStreamTrack::Destroy() {
-  mReadyState = MediaStreamTrackState::Ended;
-  if (mSource) {
-    mSource->UnregisterSink(mSink.get());
-  }
-  if (mMSGListener) {
-    if (GetOwnedStream()) {
-      RemoveListener(mMSGListener);
-    }
-    mMSGListener = nullptr;
-  }
+  SetReadyState(MediaStreamTrackState::Ended);
   // Remove all listeners -- avoid iterating over the list we're removing from
   const nsTArray<RefPtr<MediaStreamTrackListener>> trackListeners(
       mTrackListeners);
@@ -240,18 +241,16 @@
 NS_IMPL_CYCLE_COLLECTION_UNLINK_BEGIN_INHERITED(MediaStreamTrack,
                                                 DOMEventTargetHelper)
   tmp->Destroy();
-  NS_IMPL_CYCLE_COLLECTION_UNLINK(mOwningStream)
+  NS_IMPL_CYCLE_COLLECTION_UNLINK(mWindow)
   NS_IMPL_CYCLE_COLLECTION_UNLINK(mSource)
-  NS_IMPL_CYCLE_COLLECTION_UNLINK(mOriginalTrack)
   NS_IMPL_CYCLE_COLLECTION_UNLINK(mPrincipal)
   NS_IMPL_CYCLE_COLLECTION_UNLINK(mPendingPrincipal)
 NS_IMPL_CYCLE_COLLECTION_UNLINK_END
 
 NS_IMPL_CYCLE_COLLECTION_TRAVERSE_BEGIN_INHERITED(MediaStreamTrack,
                                                   DOMEventTargetHelper)
-  NS_IMPL_CYCLE_COLLECTION_TRAVERSE(mOwningStream)
+  NS_IMPL_CYCLE_COLLECTION_TRAVERSE(mWindow)
   NS_IMPL_CYCLE_COLLECTION_TRAVERSE(mSource)
-  NS_IMPL_CYCLE_COLLECTION_TRAVERSE(mOriginalTrack)
   NS_IMPL_CYCLE_COLLECTION_TRAVERSE(mPrincipal)
   NS_IMPL_CYCLE_COLLECTION_TRAVERSE(mPendingPrincipal)
 NS_IMPL_CYCLE_COLLECTION_TRAVERSE_END
@@ -261,11 +260,6 @@
 NS_INTERFACE_MAP_BEGIN_CYCLE_COLLECTION(MediaStreamTrack)
 NS_INTERFACE_MAP_END_INHERITING(DOMEventTargetHelper)
 
-nsPIDOMWindowInner* MediaStreamTrack::GetParentObject() const {
-  MOZ_RELEASE_ASSERT(mOwningStream);
-  return mOwningStream->GetParentObject();
-}
-
 JSObject* MediaStreamTrack::WrapObject(JSContext* aCx,
                                        JS::Handle<JSObject*> aGivenProto) {
   return MediaStreamTrack_Binding::Wrap(aCx, this, aGivenProto);
@@ -282,9 +276,14 @@
   }
 
   mEnabled = aEnabled;
-  GetOwnedStream()->SetTrackEnabled(
-      mTrackID,
-      mEnabled ? DisabledTrackMode::ENABLED : DisabledTrackMode::SILENCE_BLACK);
+
+  if (Ended()) {
+    return;
+  }
+
+  mStream->SetTrackEnabled(mTrackID, mEnabled
+                                         ? DisabledTrackMode::ENABLED
+                                         : DisabledTrackMode::SILENCE_BLACK);
   GetSource().SinkEnabledStateChanged();
 }
 
@@ -296,21 +295,7 @@
     return;
   }
 
-  if (!mSource) {
-    MOZ_ASSERT(false);
-    return;
-  }
-
-  mSource->UnregisterSink(mSink.get());
-
-  MOZ_ASSERT(mOwningStream,
-             "Every MediaStreamTrack needs an owning DOMMediaStream");
-  DOMMediaStream::TrackPort* port = mOwningStream->FindOwnedTrackPort(*this);
-  MOZ_ASSERT(port,
-             "A MediaStreamTrack must exist in its owning DOMMediaStream");
-  Unused << port->BlockSourceTrackId(mInputTrackID, BlockingMode::CREATION);
-
-  mReadyState = MediaStreamTrackState::Ended;
+  SetReadyState(MediaStreamTrackState::Ended);
 
   NotifyEnded();
 }
@@ -346,8 +331,7 @@
                          this, NS_ConvertUTF16toUTF8(str).get()));
   }
 
-  nsPIDOMWindowInner* window = mOwningStream->GetParentObject();
-  nsIGlobalObject* go = window ? window->AsGlobal() : nullptr;
+  nsIGlobalObject* go = mWindow ? mWindow->AsGlobal() : nullptr;
 
   RefPtr<Promise> promise = Promise::Create(go, aRv);
   if (aRv.Failed()) {
@@ -367,29 +351,28 @@
       ->Then(
           GetCurrentThreadSerialEventTarget(), __func__,
           [this, self, promise, aConstraints](bool aDummy) {
-            nsPIDOMWindowInner* window = mOwningStream->GetParentObject();
-            if (!window || !window->IsCurrentInnerWindow()) {
+            if (!mWindow || !mWindow->IsCurrentInnerWindow()) {
               return;  // Leave Promise pending after navigation by design.
             }
             mConstraints = aConstraints;
             promise->MaybeResolve(false);
           },
           [this, self, promise](const RefPtr<MediaMgrError>& aError) {
-            nsPIDOMWindowInner* window = mOwningStream->GetParentObject();
-            if (!window || !window->IsCurrentInnerWindow()) {
+            if (!mWindow || !mWindow->IsCurrentInnerWindow()) {
               return;  // Leave Promise pending after navigation by design.
             }
-            promise->MaybeReject(MakeRefPtr<MediaStreamError>(window, *aError));
+            promise->MaybeReject(
+                MakeRefPtr<MediaStreamError>(mWindow, *aError));
           });
   return promise.forget();
 }
 
-MediaStreamGraph* MediaStreamTrack::Graph() {
-  return GetOwnedStream()->Graph();
-}
+ProcessedMediaStream* MediaStreamTrack::GetStream() const { return mStream; }
 
-MediaStreamGraphImpl* MediaStreamTrack::GraphImpl() {
-  return GetOwnedStream()->GraphImpl();
+MediaStreamGraph* MediaStreamTrack::Graph() const { return mStream->Graph(); }
+
+MediaStreamGraphImpl* MediaStreamTrack::GraphImpl() const {
+  return mStream->GraphImpl();
 }
 
 void MediaStreamTrack::SetPrincipal(nsIPrincipal* aPrincipal) {
@@ -506,16 +489,11 @@
 }
 
 already_AddRefed<MediaStreamTrack> MediaStreamTrack::Clone() {
-  // MediaStreamTracks are currently governed by streams, so we need a dummy
-  // DOMMediaStream to own our track clone.
-  RefPtr<DOMMediaStream> newStream =
-      new DOMMediaStream(mOwningStream->GetParentObject());
-
-  MediaStreamGraph* graph = Graph();
-  newStream->InitOwnedStreamCommon(graph);
-  newStream->InitPlaybackStreamCommon(graph);
-
-  return newStream->CloneDOMTrack(*this, mTrackID);
+  RefPtr<MediaStreamTrack> newTrack = CloneInternal();
+  newTrack->SetEnabled(Enabled());
+  newTrack->SetMuted(Muted());
+  MOZ_DIAGNOSTIC_ASSERT(newTrack->ReadyState() == ReadyState());
+  return newTrack.forget();
 }
 
 void MediaStreamTrack::SetReadyState(MediaStreamTrackState aState) {
@@ -523,9 +501,27 @@
                aState == MediaStreamTrackState::Live),
              "We don't support overriding the ready state from ended to live");
 
+  if (Ended()) {
+    return;
+  }
+
   if (mReadyState == MediaStreamTrackState::Live &&
-      aState == MediaStreamTrackState::Ended && mSource) {
-    mSource->UnregisterSink(mSink.get());
+      aState == MediaStreamTrackState::Ended) {
+    if (mSource) {
+      mSource->UnregisterSink(mSink.get());
+    }
+    if (mMSGListener) {
+      RemoveListener(mMSGListener);
+      mMSGListener = nullptr;
+    }
+    if (mPort) {
+      mPort->Destroy();
+      mPort = nullptr;
+    }
+    if (mStream) {
+      mStream->Destroy();
+      mStream = nullptr;
+    }
   }
 
   mReadyState = aState;
@@ -540,63 +536,33 @@
 
   LOG(LogLevel::Info, ("MediaStreamTrack %p ended", this));
 
-  if (!mSource) {
-    MOZ_ASSERT(false);
-    return;
-  }
-
-  mSource->UnregisterSink(mSink.get());
-
-  if (mMSGListener) {
-    RemoveListener(mMSGListener);
-  }
-  mMSGListener = nullptr;
-
-  mReadyState = MediaStreamTrackState::Ended;
+  SetReadyState(MediaStreamTrackState::Ended);
 
   NotifyEnded();
 
   DispatchTrustedEvent(NS_LITERAL_STRING("ended"));
 }
 
-DOMMediaStream* MediaStreamTrack::GetInputDOMStream() {
-  MediaStreamTrack* originalTrack =
-      mOriginalTrack ? mOriginalTrack.get() : this;
-  MOZ_RELEASE_ASSERT(originalTrack->mOwningStream);
-  return originalTrack->mOwningStream;
-}
-
-MediaStream* MediaStreamTrack::GetInputStream() {
-  DOMMediaStream* inputDOMStream = GetInputDOMStream();
-  MOZ_RELEASE_ASSERT(inputDOMStream->GetInputStream());
-  return inputDOMStream->GetInputStream();
-}
-
-ProcessedMediaStream* MediaStreamTrack::GetOwnedStream() {
-  if (!mOwningStream) {
-    return nullptr;
-  }
-
-  return mOwningStream->GetOwnedStream();
-}
-
 void MediaStreamTrack::AddListener(MediaStreamTrackListener* aListener) {
   LOG(LogLevel::Debug,
       ("MediaStreamTrack %p adding listener %p", this, aListener));
-  MOZ_ASSERT(GetOwnedStream());
+  mTrackListeners.AppendElement(aListener);
 
-  GetOwnedStream()->AddTrackListener(aListener, mTrackID);
-  mTrackListeners.AppendElement(aListener);
+  if (Ended()) {
+    return;
+  }
+  mStream->AddTrackListener(aListener, mTrackID);
 }
 
 void MediaStreamTrack::RemoveListener(MediaStreamTrackListener* aListener) {
   LOG(LogLevel::Debug,
       ("MediaStreamTrack %p removing listener %p", this, aListener));
+  mTrackListeners.RemoveElement(aListener);
 
-  if (GetOwnedStream()) {
-    GetOwnedStream()->RemoveTrackListener(aListener, mTrackID);
-    mTrackListeners.RemoveElement(aListener);
+  if (Ended()) {
+    return;
   }
+  mStream->RemoveTrackListener(aListener, mTrackID);
 }
 
 void MediaStreamTrack::AddDirectListener(
@@ -604,43 +570,36 @@
   LOG(LogLevel::Debug, ("MediaStreamTrack %p (%s) adding direct listener %p to "
                         "stream %p, track %d",
                         this, AsAudioStreamTrack() ? "audio" : "video",
-                        aListener, GetOwnedStream(), mTrackID));
-  MOZ_ASSERT(GetOwnedStream());
+                        aListener, mStream.get(), mTrackID));
+  mDirectTrackListeners.AppendElement(aListener);
 
-  GetOwnedStream()->AddDirectTrackListener(aListener, mTrackID);
-  mDirectTrackListeners.AppendElement(aListener);
+  if (Ended()) {
+    return;
+  }
+  mStream->AddDirectTrackListener(aListener, mTrackID);
 }
 
 void MediaStreamTrack::RemoveDirectListener(
     DirectMediaStreamTrackListener* aListener) {
   LOG(LogLevel::Debug,
       ("MediaStreamTrack %p removing direct listener %p from stream %p", this,
-       aListener, GetOwnedStream()));
+       aListener, mStream.get()));
+  mDirectTrackListeners.RemoveElement(aListener);
 
-  if (GetOwnedStream()) {
-    GetOwnedStream()->RemoveDirectTrackListener(aListener, mTrackID);
-    mDirectTrackListeners.RemoveElement(aListener);
+  if (Ended()) {
+    return;
   }
+  mStream->RemoveDirectTrackListener(aListener, mTrackID);
 }
 
 already_AddRefed<MediaInputPort> MediaStreamTrack::ForwardTrackContentsTo(
-    ProcessedMediaStream* aStream, TrackID aDestinationTrackID) {
+    ProcessedMediaStream* aStream) {
   MOZ_ASSERT(NS_IsMainThread());
   MOZ_RELEASE_ASSERT(aStream);
-  RefPtr<MediaInputPort> port = aStream->AllocateInputPort(
-      GetOwnedStream(), mTrackID, aDestinationTrackID);
+  RefPtr<MediaInputPort> port =
+      aStream->AllocateInputPort(mStream, mTrackID, mTrackID);
   return port.forget();
 }
 
-bool MediaStreamTrack::IsForwardedThrough(MediaInputPort* aPort) {
-  MOZ_ASSERT(NS_IsMainThread());
-  MOZ_ASSERT(aPort);
-  if (!aPort) {
-    return false;
-  }
-  return aPort->GetSource() == GetOwnedStream() &&
-         aPort->PassTrackThrough(mTrackID);
-}
-
 }  // namespace dom
 }  // namespace mozilla